src/write/stream.rs

```diff
@@ -25,35 +26,35 @@ pub async fn write_stream<'a, W, S, E>(
 ) -> Result<u64>
 where
     W: Write,
-    S: Stream<Item = std::result::Result<RowGroupIter<'a, E>, E>>,
+    F: Future<Output = std::result::Result<RowGroupIter<'a, E>, E>>,
+    S: Stream<Item = F>,
     ParquetError: From<E>,
     E: std::error::Error,
 {
     let mut offset = start_file(writer)? as u64;
 
-    let row_groups = row_groups
-        .map(|row_group| {
-            let (group, size) = write_row_group(
-                writer,
-                offset,
-                schema.columns(),
-                options.compression,
-                row_group?,
-            )?;
-            offset += size;
-            Result::Ok(group)
-        })
-        .try_collect::<Vec<_>>()
-        .await?;
+    let mut groups = vec![];
+    pin_mut!(row_groups);
+    while let Some(row_group) = row_groups.next().await {
+        let (group, size) = write_row_group(
+            writer,
+            offset,
+            schema.columns(),
+            options.compression,
+            row_group.await?,
```

Comment:
this is the main change: the row group is now a future which we need to `await`.